### ISSUE TITLE
URL version without the date

### ISSUE THREAD

**Issue by FlorianApplicam** [Microsoft/aspnet-api-versioning#220]
Hello @commonsensesoftware,

Love what you did in this projet. 
However, I have a small question that's been bothering me for some time.

Here is the context. I have a class (VersionManager) that has a static list of all the version my company has for the API. For each version, we provide all the possible values (major and minor, but also the groupVersion). This is great because using swagger, we can add the "release date" of each version. A version without a date would have an alpha / beta status.

However, now I have an issue with the routing.
I'm using ASP.NET Core, and I'm trying to see which method is best for my use-cases. For now, we're probably going to use the URL versioning.
However, the fact is that I can't not use the date in my URL. Everytime I try testing (using Swashbuckle), the URL I have to call is api/v2017-12-02.2.0/<controller> while what I want to be calling is api/v2.0/<controller>. 
Basically, a caller should only use the major and minor version numbers in the URL and not the groupVersion. The groupVersion is only an added benefit for the documentation.

I probably missed somewhere the way to do this, but I couldn't find it.
Can you please help?

Thanks.

**Comment by commonsensesoftware** [comment]
While using all of the version components together is a supported scenario, the implementation is more of a _union_ between types. Typically, service authors choose to use the group version (e.g. date) or the major/minor scheme.  No part of the API version is _optional_. What you define the version to be is what the version is.

I'm not fond of the URL segment approach, but it's nevertheless very popular. One of the consequences of the URL segment approach is that prerelease versions in the URL are very unaesthetic to the eye. Even a basic scenario `api/v2.0-beta1/foo` looks pretty _ugly_. Using any other method is more visually appealing `api/foo?api-version=2.0-beta1`.  The main reason that I favor the query string or media type method is that the URL is stable over time and version. The URL _is_ the identifier of the resource. A new version should just mean a new _representation_. (e.g. `api/v1/order/1` and `api/v2/order/1` both refer to the exact same order in the system, but may have slightly different representations)

The _status_ component of the API version is what actually confers the meaning of _prerelease_.  The normal value is meant to be something like _alpha_, _beta_, and so on. The first character must be an alpha character, so if you want to use something like a date, it would have to look something like `api/v2.0-D20171201/foo`. Once you exit the prerelease phase, it will simply be `api/v2/foo`.  The [version format](../wiki/Version-Format) will provide additional context on the format parts and their meaning.

I hope that helps clarify things and assists you in the decision-making process for your API versioning strategy.

**Comment by FlorianApplicam** [comment]
Thanks for the reply.

From what you say, it seems that what I want to do isn't possible. It seems weird that formatting the url to follow a pattern isn't possible (ie : have `api/v2.0/foo` regardless of date and status). 
To use your example, `api/v2.0-beta1/foo` would actually be called `api/v2.0/foo`. Whether it is in alpha, beta, etc... it would always stay the same.

I can understand why you prefer using the query string. Although I'm not a fan, it does indeed allow to always keep the same url regardless of the version.

For now, I'm still in the decision making phase. I'm weighing all the options to choose the one that suits me best.

**Comment by commonsensesoftware** [comment]
Fair enough. There are a quite a few alternatives. I've seen teams keep their own internal mapping, but it seems that's exactly the problem you are trying to avoid.

One thing to keep in mind is that the _prerelease_ is formally part of the version. There's no _loosey-goosey_ stuff happening. Despite the URL being _ugly_ the _prerelease_ component is part of the API version. This keeps things clear for clients. They'll never be in a situation were they use v2.0 and are wondering if it's prerelease or not. Furthermore, prereleases usually go away eventually. The official version (in your case) will always be clean (e.g. v2.0, v3.0, etc).

With respect to formatting, API versioning isn't making any assumptions. In terms of how to dispatch requests, it's extracting the API version to make the decision. On the server side, formatting makes sense and can be done. On the client side though, it makes no sense. If the client asks for `v2.0`, that's what they get. If they ask for `v2.0-beta1`, that's what they get. Things should be unambiguous to the client.

Visually, this more of what API versioning is looking at:

```
GET api/v{apiVersion}/foo?api-version={apiVersion} HTTP/1.1
Host: localhost
Accept: application/json;v={apiVersion}
Api-Version: {apiVersion}
```
Each of the strategies describe where and how to extract the value. The value is the value. There is no special manipulation or transformation.

It might be worth noting that you _can_ use multiple API versioning methods simultaneously. In fact, the default configuration supports both the query string and URL segment methods because when, where, and how they extract the API version are mutually inclusive without any extra overhead. This _could_ allow you to have something like `api/prerelease/foo?api-version=2017-12-01`. The official release version can then be `api/v2.0/foo`. This would allow you to separate release and prerelease scenarios. It doesn't solve your desire for self-documenting the date and mixes an undesirable style of versioning, but perhaps that is acceptable for prereleases. Regardless, hopefully that provides some additional ideas about how you might be able to make things work for you.

**Comment by commonsensesoftware** [comment]
I suspect that got the answers you needed to help you make your decision. Feel free to come back or reopen this issue if you have more questions.
